src/storage/src/source/postgres.rs

```diff
@@ -246,28 +248,58 @@ impl SourceConnectionBuilder for PostgresSourceConnection {
 
         if active_read_worker {
             let mut source_tables = HashMap::new();
-            let tables_iter = self.details.tables.iter();
-            for (i, (desc, casts)) in tables_iter.zip(self.table_casts).enumerate() {
-                let source_table = SourceTable {
-                    output_index: i + 1,
-                    desc: desc.clone(),
-                    casts,
-                };
-                source_tables.insert(desc.oid, source_table);
+            let tables_iter = self.publication_details.tables.iter();
+
+            for (i, desc) in tables_iter.enumerate() {
+                let output_index = i + 1;
+                // We maintain descriptions for all tables in the publication,
+                // but only casts for those we aim to use (and have validated
+                // that their types are ingestable). This also prevents us from
+                // creating snapshots for tables in the publication that are
+                // not referenced in the source.
+                match self.table_casts.get(&output_index) {
+                    Some(casts) => {
+                        let mut desc = desc.clone();
+
+                        // Rewrite all `self.text_cols` columns to text. This
+                        // is the expected output type generated by `casts`
+                        // done in planning, and this must be done here so as
+                        // to not overwrite the "native" values snapshot in
+                        // `self.publication_details`.
+                        if let Some(columns) = self.text_cols.get(&Oid(desc.oid)) {
```

Comment:
Why do we need to alter the description here? I think we should be able to remove the `text_cols` field from `PostgresSourceConnection` entirely because it already contains correctly picked cast expressions selected by the planner

Reply:
I had thought that the column descriptions were used to express how to decode the specific columns from the PG source; hadn't put together that logic is entirely determined by `column_casts`. I'm actually able to remove all of the `text_cols` logic from storage--thanks for pointing this out.